WIP gh-103741 - PEP 713 - Callable modules

amyreese opened the pull request:
Draft implementation of PEP 713:

- [x] Implement ability to call modules with `__call__` property.
- [x] Unit tests covering behavior for callable and non-callable modules
- [ ] Documentation changes
- [ ] News entry

Adds a method wrapper for module objects that makes them callable.
When called, checks for a `__call__` object on the module and calls
that instead, otherwise raises type error declaring the module
is not callable.

Adds a check in `module_getattro` to prevent returning the method
wrapper when looking for `__call__`, to prevent circular lookups
when calling the module and checking for `__call__` set by module
authors.

Adds `PyModule_Callable` helper function to determine if a module
is callable, by looking for a `__call__` property and calling the
normal `PyCallable_Check` on that object, if it exists.

Updates `PyCallable_Check` to use the new `PyModule_Callable` helper
when checking module objects, to correctly look for a callable
`__call__` property rather than the normal `tp_call` attribute on the
module's type struct.

Tested invariants:

- For modules without `__call__` (eg, stdlib):
  - `callable(mod)` -> `False`
  - `mod.__call__` -> `AttributeError`
  - `mod()` -> `TypeError`

- For module with `__call__` or `__getattr__` returning `__call__` object
  - `callable(mod)` -> `callable(mod.__call__)`
  - `mod.__call__` -> real object
  - `mod()` -> `mod.__call__()`

Setting (or unsetting) `__call__` on existing module objects makes
them callable (or uncallable) accordingly, without changing the type
or class of the module object.


<!-- gh-issue-number: gh-103741 -->
* Issue: gh-103741
<!-- /gh-issue-number -->

amyreese committed:
PEP 713: make modules callable

When making calls to core types, if no `tp_call` value is found (the
type isn't default callable) and the type in question is a `ModuleType`,
then `_PyObject_MakeTpCall` will use getattr to check for a `__call__`
attribute on the module object, and attempt to call that instead.

When checking if an object is callable, the initial `tp_call` lookup
doesn't find a value, and the object type is a module, a similar getattr
for `__call__` is made, and a callable check is made against that value
instead.

This is implemented this way, rather than by adding a `tp_call` method
directly to the base module type, as not all modules will be callable,
and the existence of a method wrapper makes it difficult to determine
if the module in question is actually callable, or just has a method
wrapper. This eliminates the need for isinstance checks and special
error handling in said method wrapper to make non-callable modules
continue to appear as non-callable objects.

amyreese committed:
Tests validating callable module implementation

Fidget-Spinner commented on Objects/call.c:
```diff
@@ -181,6 +181,13 @@ _PyObject_MakeTpCall(PyThreadState *tstate, PyObject *callable,
+            PyObject* mod_call = PyObject_GetAttr(callable, &_Py_ID(__call__));
+            if (mod_call != NULL) {
+                return _PyObject_MakeTpCall(tstate, mod_call, args, nargs, keywords);
+            }
```
I don't think you need to modify `_PyObject_MakeTpCall`. The general convention is to add something to a type object's `tp_call` or `tp_vectorcall_offset` slot. In this case, you would want to modify `Object/moduleobject.c`'s `PyModule_Type` to have a simple wrapper function. The wrapper function would do the same as you're doing now (looking up `__call__`).

amyreese replied:
I was originally going down this path, but seems to result in the undesired situation where it marks *all* modules as callable, similar to functions/methods/types, and every module object then has a default `__call__` property. That also has the effect of making `callable(module)` always true, even if `module()` immediately raises `TypeError: 'module' object is not callable`.

Quoting my commit message for 524613cdfbe33c1ad5d1197b61786f1960656f5f:

> This is implemented this way, rather than by adding a `tp_call` method
directly to the base module type, as not all modules will be callable,
and the existence of a method wrapper makes it difficult to determine
if the module in question is actually callable, or just has a method
wrapper. This eliminates the need for isinstance checks and special
error handling in said method wrapper to make non-callable modules
continue to appear as non-callable objects.

Is there a better way to do this that doesn't have this downside?

Fidget-Spinner replied:
In that case, you can add the tp_call, but special-case module types in `callable`, with the same checks you're doing now.

Fidget-Spinner commented on Objects/object.c:
```diff
@@ -1668,7 +1668,13 @@ PyCallable_Check(PyObject *x)
+        PyObject* call = PyObject_GetAttr(x, &_Py_ID(__call__));
+        PyErr_Clear();
+        return PyCallable_Check(call);
+    }
```
Ditto as above. I don't think this should be modified.

erlend-aasland replied:
Also, there's a missing return at the end of this function.

Fidget-Spinner reviewed (commented):
Thanks for the effort in writing this PEP. I have some minor comments below.

Fidget-Spinner commented on Objects/object.c:
```diff
@@ -1668,7 +1668,13 @@ PyCallable_Check(PyObject *x)
+        return 1;
+    if (Py_TYPE(x) == &PyModule_Type) {
+        PyObject* call = PyObject_GetAttr(x, &_Py_ID(__call__));
+        PyErr_Clear();
```
Just a note in case you decide to re-use this code. You should check that ``call == null && PyErr_ExceptionMatches(PyExc_AttributeError)`` before clearing the error. Otherwise valid errors (such as out of memory errors) might get cleared too.

sunmy2019 commented on Lib/test/test_call.py:
```diff
@@ -945,5 +945,27 @@ def test_function_with_many_args(self):
+        import sys
+
+        self.assertFalse(callable(sys))
+        self.assertRaisesRegex(TypeError, "'module' object is not callable", sys)
```
Some 'module' objects are now callable, maybe we can use `f"{module} object is not callable"`?

JelleZijlstra replied:
I would suggest saying something like "Module 'os' is not callable".

ambv committed:
Merge branch 'main' into callable-modules

amyreese committed:
Change how modules are callable

Adds a method wrapper for module objects that makes them callable.
When called, checks for a __call__ object on the module and calls
that instead, otherwise raises type error declaring the module
is not callable.

Adds a check in module_getattro to prevent returning the method
wrapper when looking for __call__, to prevent circular lookups
when calling the module and checking for __call__ set by module
authors.

Adds PyModule_Callable helper function to determine if a module
is callable, by looking for a __call__ property and calling the
normal PyCallable_Check on that object, if it exists.

Updates PyCallable_Check to use the new PyModule_Callable helper
when checking module objects, to correctly look for a callable
__call__ property rather than the normal tp_call attribute on the
module's type struct.

Tested invariants:

- For modules without __call__ (eg, stdlib):
  - callable(mod) -> False
  - mod.__call__ -> AttributeError
  - mod() -> TypeError

- For module with __call__ or __getattr__ returning __call__ object
  - callable(mod) -> callable(mod.__call__)
  - mod.__call__ -> real object
  - mod() -> mod.__call__()

Setting (or unsetting) __call__ on existing module objects makes
them callable (or uncallable) accordingly, without changing the type
or class of the module object.

Co-authored-by: Hood Chatham <roberthoodchatham@gmail.com>

erlend-aasland commented on Lib/test/test_call.py:
```diff
@@ -945,5 +945,48 @@ def test_function_with_many_args(self):
+                del sys.__call__
+
+    def test_callable_modules(self):
+        from . import callable_module_a, callable_module_b, callable_module_c
```
I'd prefer inlining these instead of polluting the `Lib/test` namespace with the added files. Alternatively, put them in a subdirectory.

erlend-aasland reviewed (commented):
You also need to add NEWS and What's New entries.

encukou commented on Objects/object.c:
```diff
@@ -1668,6 +1668,9 @@ PyCallable_Check(PyObject *x)
 {
     if (x == NULL)
         return 0;
+    if (PyModule_CheckExact(x)) {
```
This excludes subclasses of ModuleType.

encukou commented on Include/moduleobject.h:
```diff
@@ -34,6 +34,7 @@ PyAPI_FUNC(int) _PyModuleSpec_IsInitializing(PyObject *);
 #endif
 PyAPI_FUNC(PyModuleDef*) PyModule_GetDef(PyObject*);
 PyAPI_FUNC(void*) PyModule_GetState(PyObject*);
+PyAPI_FUNC(int) PyModule_Callable(PyObject *);
```
This should be in `Include/cpython/*.h` or in an `#ifndef Py_LIMITED_API` block. 
Or, if you want to add it to the limited API, see [docs in the devguide](https://devguide.python.org/developer-workflow/c-api/#limited-api).
